modules/ocf_www/files/build-vhosts

```diff
@@ -14,190 +11,350 @@ from pathlib import Path
-            ({domain} if ssl_options else set()) | set(vhost['aliases'])
-        ):
+        # if it has aliases defined, add HTTP and HTTPS domains redirects too
+        redirects = set()
+        if use_ssl:
+            redirects.add((primary.fqdn, None))
+        for alias in vhost['aliases']:
+            redirects.add((alias, None))
+            ssl = SSL(alias)
+            if ssl.is_valid:
+                redirects.add((alias, ssl))
+
+        for fqdn, ssl in redirects:
             vhosts.add(VirtualHost(
-                fqdn=redirect_domain,
-                user=vhost['username'],
-                relative_docroot='/',
+                fqdn=fqdn,
+                user=user,
+                comment='{} (redirect to {})'.format(fqdn, primary.fqdn),
+                ssl=ssl,
+                socket_name=None,
+                relative_docroot=None,
                 redirect_dest=primary.canonical_url,
-                ssl_cert=None,
-                ssl_key=None,
-                ssl_chain=None,
-                use_hsts=None,
-                disabled=False,
             ))
 
     return '\n\n'.join(
-        str(vhost)
+        template.render(vhost=vhost)
         for vhost in sorted(
             vhosts,
-            key=lambda vhost: (vhost.user, vhost.fqdn, vhost.port),
+            key=lambda vhost: (vhost.user, vhost.fqdn, bool(vhost.ssl)),
         )
-    ) + '\n'
+    )
+
 
+def test_and_overwrite_config(config_path, new_config):
+    """Diffs and tests the new config and overwrites the old config if
+    the test passes.
+
+    Returns whether the old config was overwritten.
+    """
+    (_, new_path), (__, old_path) = (
+        tempfile.mkstemp(
+            # Prefix with '.' so temp files don't get included
+            prefix='.' + os.path.basename(config_path),
+            dir=os.path.dirname(config_path),
+        )
+        for i in range(2)
+    )
+    try:
+        os.chmod(new_path, 0o644)
+        with open(new_path, 'w') as f:
+            f.write(new_config)
+        report('Wrote to ' + new_path)
 
-def main(argv=None):
+        report('Diff between current and new file:')
+        ret = subprocess.call(('diff', '-u', config_path, new_path))
+        report('[end diff]')
+
+        if not ret:
+            report('Nothing changed, not doing anything.')
+            return False
+
+        # Save existing config
+        shutil.copyfile(config_path, old_path)
+        report('Copied existing config to ' + old_path)
+
+        # Overwrite with new config and test
+        os.rename(new_path, config_path)
+
+        report('Performing config test.')
+        if target == 'web':
+            ret = subprocess.call(('apachectl', 'configtest'))
+        else:
+            # nginx is moronic and prints "test successful" on stderr,
+            # so echo to stderr only if nginx -t failed
+            p = subprocess.Popen(('nginx', '-t'), stderr=subprocess.PIPE)
+            ret = p.wait()
```

Comment:
this can deadlock if nginx produces too much output (it fills the output buffer but nothing is reading from it yet; python is waiting for the process to exit, which it never will because it's trying to write output)

instead you probably want something like...

```python
stdout, stderr = p.communicate()
ret = p.returncode  # communicate() does a wait(), so you don't need to do it again
```